trace_decoder/src/lib.rs

```diff
@@ -109,30 +63,402 @@
+/// rely on a separate EVM to run the txn and supply this data for us.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct TxnTrace {
+    /// If the balance changed, then the new balance will appear here. Will be
+    /// `None` if no change.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub balance: Option<U256>,
+
+    /// If the nonce changed, then the new nonce will appear here. Will be
+    /// `None` if no change.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub nonce: Option<U256>,
+
+    /// Account addresses that were only read by the txn.
+    ///
+    /// Note that if storage is written to, then it does not need to appear in
+    /// this list (but is also fine if it does).
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub storage_read: Option<Vec<H256>>,
+
+    /// Account storage addresses that were mutated by the txn along with their
+    /// new value.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub storage_written: Option<HashMap<H256, U256>>,
+
+    /// Contract code that this address accessed.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub code_usage: Option<ContractCodeUsage>,
+
+    /// True if the account existed before this txn but self-destructed at the
+    /// end of this txn.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub self_destructed: Option<bool>,
+}
+
+/// Contract code access type. Used by txn traces.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum ContractCodeUsage {
+    /// Contract was read.
+    Read(H256),
+
+    /// Contract was created (and these are the bytes). Note that this new
+    /// contract code will not appear in the [`BlockTrace`] map.
+    Write(#[serde(with = "crate::hex")] Vec<u8>),
+}
+
+impl ContractCodeUsage {
+    fn get_code_hash(&self) -> H256 {
+        match self {
+            ContractCodeUsage::Read(hash) => *hash,
+            ContractCodeUsage::Write(bytes) => hash(bytes),
+        }
+    }
+}
+
+/// Other data that is needed for proof gen.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct OtherBlockData {
+    /// Data that is specific to the block.
+    pub b_data: BlockLevelData,
+    /// State trie root hash at the checkpoint.
+    pub checkpoint_state_trie_root: H256,
+}
+
+/// Data that is specific to a block and is constant for all txns in a given
+/// block.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct BlockLevelData {
+    /// All block data excluding block hashes and withdrawals.
+    pub b_meta: BlockMetadata,
+    /// Block hashes: the previous 256 block hashes and the current block hash.
+    pub b_hashes: BlockHashes,
+    /// Block withdrawal addresses and values.
+    pub withdrawals: Vec<(Address, U256)>,
+}
+
+/// TODO(0xaatif): https://github.com/0xPolygonZero/zk_evm/issues/275
+///                document this once we have the API finalized
+pub fn entrypoint(
```

Comment:
Non-blocking: As this will be the only API function of the trace decoder I wonder if some other name would be more appropriate, e.g. `process`, or `generate_inputs`?

Reply:
> I'd prefer this PR didn't bikeshed names - we can't finalize an API until we see what the SMT backend requires.

From the review description :)